frontend/src/frontend_benchmark/source_generator.py

```diff
@@ -0,0 +1,185 @@
+            f.write(template)
+
+    def _build_main_template(self) -> str:
+        header = self.get_header_import_string()
+        vars_def = self.callgraph.format_vars_definition()
+        function_call = self.callgraph.function_call_signature_for(self.callgraph.entry_point)
+        variable = 'loops'
+        arg_template = self._build_arg_template(variable)
+        template = (
+            f'#include <unistd.h>\n'
+            f'#include <stdio.h>\n'
+            f'#include <stdlib.h>\n'
+            f'{header}\n\n'
+            f'{vars_def}\n'
+            'int main(int argc, char **argv) {\n'
+            f'unsigned long {variable} = 1;\n'
+            f'{arg_template}\n'
+            f'for (int i = 0; i < {variable}; i++)'
+            ' {\n'
+            f'{function_call}();\n'
+            '}\n'
+            '}\n'
+        )
+        return template
+
+    def get_header_import_string(self) -> str:
+        return f'#include "{self.header_file}"'
+
+    def _build_arg_template(self, variable) -> str:
+        template = (
+            'int c;\n'
+            'while ((c = getopt(argc, argv, "l:")) != -1) {\n'
+            'switch (c) {\n'
+            "case 'l':\n"
+            f'{variable} = strtoul(optarg, NULL, 0);\n'
+            '}\n'
+            '}'
+        )
+        return template
+
+    def append_path_to_output_dir(self, path) -> str:
+        return os.path.join(self.output_dir, path)
+
+    def write_headers(self) -> None:
+        header_path = self.append_path_to_output_dir(self.header_file)
+        with open(header_path, 'w') as f:
+            f.write(self.callgraph.format_vars_declaration())
+            f.write(self.callgraph.format_headers())
+
+    def write_functions(self, splits: int = None) -> Collection[str]:
+        grouper = FileFunctionMapper(self.callgraph)
+        file_to_function_name_mapping = grouper.group_functions(splits)
+        for function_file, functions in file_to_function_name_mapping.items():
+            self.write_header_import(function_file)
+            for function_name in functions:
+                self.write_function(function_name, function_file)
+        return file_to_function_name_mapping.keys()
+
+    def write_header_import(self, path) -> None:
+        path = self.append_path_to_output_dir(path)
+        header = self.get_header_import_string()
+        with open(path, 'a') as f:
+            f.write(f'{header}\n\n')
+
+    def write_function(self, function_name: int, path: str) -> None:
+        path = self.append_path_to_output_dir(path)
+        string = self.callgraph.format_function(function_name)
+        with open(path, 'a') as f:
+            f.write(string + '\n')
+
+    def write_makefile(self, function_files: Collection[str]) -> None:
+        makefile_path = self.append_path_to_output_dir('Makefile')
+        with open(makefile_path, 'w') as f:
+            dependencies = self.get_object_files_to_c_files_mapping(function_files)
```

Comment:
Being pedantic like this is okay, but would it work if we wrote a Makefile that finds all *.c files in the directory, builds *.o out of all of them with a single rule, then links all of them together? Then we don't need to regenerate the Makefile every time. In fact, we might even be able to just provide one up front.

Reply:
Good idea. We can definitely do create a single rule to compile all the c files.

re providing a Makefile upfront: We can do that, but I can see us wanting variable compilation options in the future. It might be easier to add that functionality with the current code.